src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -886,12 +922,54 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
 
                         if (status.ErrorCode == SecurityStatusPalErrorCode.ContextExpired)
                         {
-                            return 0;
+                            receivedEOF = true;
+                            break;
                         }
 
                         throw new IOException(SR.net_io_decrypt, SslStreamPal.GetException(status));
                     }
+
+                    if (_decryptedBytesCount == 0)
+                    {
+                        // This may be service frame to update keys or renegotiation.
+                        // Keep waiting for real data.
+                        if (processedLength > 0)
+                        {
+                            // We have some data. Hand them to caller before going for another iteration.
+                            break;
+                        }
+
+                        continue;
+                    }
+
+                    // This will either copy data from rented buffer or adjust final buffer as needed.
+                    // In both cases _decryptedBytesOffset and _decryptedBytesCount will be updated as needed.
+                    processedLength += CopyDecryptedData(buffer);
+
+                    if (_decryptedBytesCount > 0)
+                    {
+                        // We have more than enough to fill provided buffer.
+                        break;
+                    }
+
+                    // Check if we have enough data to process next TLS frame
+                    if (_internalBufferCount > SecureChannel.ReadHeaderSize && _framing == Framing.SinceSSL3)
```

Comment:
I don't understand why this logic seems to be rather different than the logic above that checks whether we have a full frame or not.

Maybe we should have a helper function like, bool IsFullFrameBuffered(out frameSize)?

Reply:
I'll check if helper function would make sense. The difference is that if we have less than ReadHeaderSize bytes at the beginning we would try to read it so we can determine size of the frame. In this case we simply bail out without trying. The _framing check is probably no necessary. We don't have much coverage for old protocol versions so I wanted to enable the optimization only for TLS where I have more confidence. 
Let me know if that make sense.